worker/codalabworker/aws_batch.py

```diff
@@ -0,0 +1,539 @@
+                        queue_name=self._queue_name,
+                        worker=self._worker,
+                        bundle_service=self._bundle_service,
+                        bundle_path=self._bundle_path,
+                        resources=self._resources,
+                        dependencies=dependencies)
+        self._fsm = fsm.ThreadedFiniteStateMachine(state, sleep_time=5.0)
+        return self._fsm
+
+
+def event(name, **payload):
+    return {
+        'name': name,
+        'payload': payload
+    }
+
+
+class BatchStatus(object):
+    """
+    Constants for the statuses a Batch job can be in.
+    see: https://docs.aws.amazon.com/batch/latest/APIReference/API_JobDetail.html
+    """
+    Submitted = 'SUBMITTED'
+    Pending = 'PENDING'
+    Runnable = 'RUNNABLE'
+    Starting = 'STARTING'
+    Running = 'RUNNING'
+    Succeeded = 'SUCCEEDED'
+    Failed = 'FAILED'
+
+
+class AwsBatchRunState(fsm.State):
+    def __init__(self, bundle, batch_client, queue_name, worker, bundle_service, bundle_path, resources, dependencies):
+        self._bundle = bundle
+        self._batch_client = batch_client
+        self._queue_name = queue_name
+        self._worker = worker
+        self._bundle_service = bundle_service
+        self._bundle_path = bundle_path
+        self._logger = logging.getLogger(self.uuid)
+        self._resources = resources
+        self._dependencies = dependencies
+
+    @property
+    def logger(self):
+        return self._logger
+
+    @property
+    def uuid(self):
+        return self._bundle['uuid']
+
+    @property
+    def batch_queue(self):
+        return self._queue_name
+
+    @property
+    def is_shared_file_system(self):
+        return self._worker.shared_file_system
+
+    @property
+    def metadata(self):
+        return self._bundle['metadata']
+
+    @property
+    def resources(self):
+        return self._resources
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    def transition(self, NewState):
+        new_state = NewState(bundle=self._bundle,
+                             batch_client=self._batch_client,
+                             queue_name=self._queue_name,
+                             worker=self._worker,
+                             bundle_service=self._bundle_service,
+                             bundle_path=self._bundle_path,
+                             resources=self.resources,
+                             dependencies=self._dependencies)
+        self.logger.info("Job %s transitioning %s -> %s", self.uuid, self.name, new_state.name)
+        return new_state
+
+    def update_metadata(self, **kwargs):
+        self.logger.debug("Updating metadata: %s", kwargs)
+        # Update the bundle locally
+        self._bundle['metadata'].update(kwargs)
+        # Update the bundle on the bundle service
+        self._bundle_service.update_bundle_metadata(self._worker.id, self._bundle['uuid'], kwargs)
```

Comment:
The `Run` code avoids blowing up if the bundle service throws an exception here. Should we replicate that behavior?